api/tests/v2/test_email.py

```diff
@@ -0,0 +1,127 @@
+from django.test import TestCase
+from rest_framework.test import APIRequestFactory, force_authenticate
+import mock
+
+from api.v2.views import (
+    VolumeSupportEmailViewSet, InstanceSupportEmailViewSet,
+    FeedbackEmailViewSet, ResourceEmailViewSet
+)
+from api.tests.factories import (
+    AnonymousUserFactory, InstanceFactory, UserFactory, VolumeFactory
+)
+
+
+class EmailTestHelper(TestCase):
+    """
+    This class is used for testing the following ViewSets:
+        VolumeSupportEmailViewSet, InstanceSupportEmailViewSet,
+        FeedbackEmailViewSet, ResourceEmailViewSet
+
+    Each of these is tested the same way by mocking an email method and
+    checking if it is called or not called. By extending this class, they
+    can all use the methods of this class to efficiently test email using
+    requests to different API endpoints containing different data
+    """
+
+    def email_test_helper(self, request, cls, mock_method):
+        view = cls.as_view({'post': 'create'})
+        force_authenticate(request, user=UserFactory())
+        with mock.patch(mock_method) as mock_email_support:
+            response = view(request)
+            mock_email_support.assert_called_once()
+            self.assertEquals(response.status_code, 200)
+
+    def anonymous_email_test_helper(self, request, cls, mock_method):
+        view = cls.as_view({'post': 'create'})
+        force_authenticate(request, user=AnonymousUserFactory())
+        with mock.patch(mock_method) as mock_email_support:
+            response = view(request)
+            mock_email_support.assert_not_called()
+            self.assertEquals(response.status_code, 403)
```

Comment:
Helper methods should not perform tests. When you're reading a test you don't want to trace down assertions in abstractions. 

A test is all about assertions. When you read a test you want to read as little setup code, and only have to make sense about the assertions. It is worth abstracting all the creation and mocking to have simple assertions, but you don't want to hide what assertions are happening in each test.

As a note on style, code duplication is not the worst sin, the worst sin is a poor abstraction. It's a common pattern to avoid code duplication at all costs, this often leads to methods that are so specific or obtuse that they could never be reused, or don't really make sense as a unit of work. By analogy, think about shell aliases and bash functions. You save a bash function when you think it may be useful later, something that is generally useful. You wouldn't make a bash function which searched /var/log/celery/<some node> for the string "cdosborn". If you did you would accrue all these useless functions. Now imagine that you were tasked with understanding someone else's set of functions written this way. It would be hard and not fun (yet ironically this is much of software development). 

Reflecting on that idea, can you appreciate my objection to  `email_test_helper`. I would struggle to come up with a sentence to describe exactly what it does. It's just a clever bundling of statements which prevents code duplication at the cost of being understandable and re-usable. 

Another test for is this abstraction good: do the arguments make sense, could I assume what arguments would need to be passed to this function? Your function `email_admin(subject, body, sender)` is a perfect example of a good abstraction, I wouldn't expect that I would ever need to look inside that function.

Reply:
Here is an example where I broke out of a ton of the duplicate code into helpful functions at the bottom: `api/tests/v2/test_resource_requests.py`.

Each test reads, blah blah blah `assert` `asert` etc.